src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -680,25 +692,138 @@ private void genIncubatorModules(ClassBuilder clb) {
          */
         private void genModuleDescriptorsMethod(ClassBuilder clb) {
+            if (moduleInfos.size() <= moduleDescriptorsPerMethod) {
+                clb.withMethodBody(
+                        "moduleDescriptors",
+                        MTD_ModuleDescriptorArray,
+                        ACC_PUBLIC,
+                        cob -> {
+                            cob.constantInstruction(moduleInfos.size())
+                               .anewarray(CD_MODULE_DESCRIPTOR)
+                               .astore(MD_VAR);
+
+                            for (int index = 0; index < moduleInfos.size(); index++) {
+                                ModuleInfo minfo = moduleInfos.get(index);
+                                new ModuleDescriptorBuilder(cob,
+                                                            minfo.descriptor(),
+                                                            minfo.packages(),
+                                                            index).build();
+                            }
+                            cob.aload(MD_VAR)
+                               .areturn();
+                        });
+                return;
+            }
+
+            // Instead of adding the module descriptor calls in this method,
+            // helper methods are created in separate helper methods.
+            // Let m1, m2, ..., mn be the calls to the module descriptor builder
+            // Here, n is larger than moduleDescriptorsPerMethod, which will hit the 64kb limit of method length.
+            // thus, we create f1, f2, ... where
+            //   - f1 calls m1, ... m_{moduleDescriptorsPerMethod-1},
+            //   - f2 calls m_{moduleDescriptorsPerMethod}, ... m_{2xmoduleDescriptorsPerMethod-1},
+            //   - etc.
+            //
+            // Inside m, the class SystemModulesClassGenerator.DedupSetBuilder is used.
+            // That class creates sets caching variables. It stores its caches in the local storage.
+            // Since the local storage is destroyed across each method, but the sets are needed across methods,
+            // the local variables are passed on to the next helper method using a list of these variables.
+
+            List<List<ModuleInfo>> splitModuleInfos = new ArrayList<>();
+            List<ModuleInfo> currentModuleInfos = null;
+            for (int index = 0; index < moduleInfos.size(); index++) {
+                if (index % moduleDescriptorsPerMethod == 0) {
+                    currentModuleInfos = new ArrayList<>();
+                    splitModuleInfos.add(currentModuleInfos);
+                }
+                currentModuleInfos.add(moduleInfos.get(index));
+            }
+
+            String helperMethodNamePrefix = "sub";
+            ClassDesc arrayListClassDesc = ClassDesc.ofInternalName("java/util/ArrayList");
+
+            int firstVar = nextLocalVar;
+            var wrapper = new Object() {
```

Comment:
I think you can simply use a local variable at the beginning of each iteration.  See below.